README
This program was inspired by discussion about Funkey and
2.6 Linux kernels. Just read what Vojtech Pavlik think
about Funkey patch. (at the end of file)

How to use this program?

1. Compile kernel with input support.
(generic kernels should have it)

1a. If compiled as a module (best choice) load module.
(type "modprobe evdev" - as root)

2. If you know what permissions are set it correctly ;-P
(e.g. prevent users from installing key loggers)

3. Compile program.
(type "./configure ; make")

4. Install program.
(type "make install" AS ROOT)

5. Run keytest and test if your keys works.
(type "keytest")

6. Run learnkeys and make config file.
(type "learnkeys config_file_name")

7. Edit config file generated by learnkeys.
(use your favorite editor)

8. Run it with the config file name.
(type "esekeyd config_file_name",
you don`t have to run it as root if you set permissions correctly,
if you want to do root need things (like init 5) use sudo,
my config file is included in exaples directory)

9. Have fun!

Krzysztof Burghardt <krzysztof@burghardt.pl>

-=[ LKML ]=---------------------------------------------------------------

Form: Vojtech Pavlik <vojtech@suse.cz>
Subject: Re:Re: [BUG] setkeycodes with 2.6.0-test5 / -test6

On Tue, Sep 30, 2003 at 10:53:08AM +0100, Ivo van Doorn wrote:
> Hi!
>

> I'va been working on a kernelpatch for the 2.6 test kernel named the
> Funkey patch. This patch has been created by Rick van Rein for 2.2 and
> 2.4 kernels: website is here: http://rck.vanrein.org/linux/funkey.

> When i ported this patch to 2.6 I run into small problem., which also
> occurs when no patch is applied. First I will give a small
> explanation on how the Funkey patch works:

> Modern keyboards have extra multimedia keys (think of the "www" "mail"
> "search" sound volume etc.) By using setkeycodes to give these buttons
> a keycode and in the keymap giving that keycode a highbyte the key
> gets picked up by the funkey patch and instead of sending the signal
> through /dev/console, it gets send through /dev/funkey. Using the
> matching daemon the /dev/funkey can be read and programs can be
> launched when the buton is pressed. "www" button can for example start
> lynx, or mozilla or whatever.

You shouldn't need any patch for 2.6 to implement this functionality.
Basically you need to modify the funkey daemon to do this:

1) Open /proc/bus/input/devices and read it to find out which device
is the system keyboard.
2) Open the correct /dev/input/event? device
3) Program the scancode/keycode mappings using the EVIOCSKEYCODE
ioctl()
4) Use the EVIOCGRAB ioctl() to prevent any other program to get
keycodes from the keyboard.
5) Open /dev/uinput and create a new virtual keyboard
6) Read events from /dev/input/event? and process them. Either forward
them to /dev/uinput (normal keys) or do whatever action is needed
(fun keys)

Does that sound reasonable?

Another option is to omit steps 4 and 5, and just let the extra keys
reach the applications. If they're not mapped in kernel/xfree keymaps,
they'll be simply ignored.

One more thing: Microsoft standardized some of the WWW/Multimedia keys.
I'll be changing the default scancode/keycode/rawmode tables to support
them soon.

[...]

-- 
Vojtech Pavlik
SuSE Labs, SuSE CR


NEWS

 Release Notes
 =============

 *******************
 1.2.7 (04 Jun 2010)
 *******************

* New feature allow to handle multiple keys sequences/combinations.
* New feature allow handle both key press and releases by Jiri Pinkava
* Fix that allow numeric keys in configuration by Jiri Pinkava
* Fix to problem when there is no newline on last line in config file by Jiri
  Pinkava

 *******************
 1.2.6 (29 Jul 2009)
 *******************

* Fixed string overflow bug reported by Karl Dahlke.

 *******************
 1.2.5 (15 Nov 2008)
 *******************

* Changed man path in doc/*.1 to FHS which is used by Debian.

 *******************
 1.2.4 (23 Sep 2008)
 *******************

* Fixed cross-compilation bug reported by hartleys.
* Removed debian directory (will try to include esekeyd in Debian).
* Removed old directory (Linux 2.4 users can use 1.2.3, but 2.4 support
  is no longer maintained).
* Removed lib directory (any modern (g)libc has getopt).

 *******************
 1.2.3 (01 Jun 2006)
 *******************

* Fixed bug reported by Евгений Вдовин.

 *******************
 1.2.2 (21 Feb 2006)
 *******************

* Debianization.
* signal handling for daemon.
* support for /var/run/esekeyd.pid file.
* added Gentoo initscript.
* added man pages.

 *******************
 1.2.1 (15 May 2004)
 *******************

* added option to override autodetection of input device (now you can bind
  esekeyd to your remote control input device, not only 1st detected keyboard).
* minor bugfix.

 *******************
 1.2.0 (24 Jan 2004)
 *******************

* most code of esekeyd rewriten.
* rewritten using autoconf, automake and autoheader.
* added two new programs keytest and learnkeys.
* added support for config files.
* removed funkey.sh (replaced by config file).
* added support for all keys form <linux/input.h>.
* minor bugfix.


src/parser.c
/* autogenerated by mkparser - DO NOT EDIT */

#include "esekey.h"

char*
parse(struct input_event ev)
{
	static char *ret_val;

	if (ev.type == EV_KEY && ev.value == 1) {

		switch (ev.code) {

			case 1:
				return "ESC";
			case 2:
				return "1";
			case 3:
				return "2";
			case 4:
				return "3";
			case 5:
				return "4";
			case 6:
				return "5";
			case 7:
				return "6";
			case 8:
				return "7";
			case 9:
				return "8";
			case 10:
				return "9";
			case 11:
				return "0";
			case 12:
				return "MINUS";
			case 13:
				return "EQUAL";
			case 14:
				return "BACKSPACE";
			case 15:
				return "TAB";
			case 16:
				return "Q";
			case 17:
				return "W";
			case 18:
				return "E";
			case 19:
				return "R";
			case 20:
				return "T";
			case 21:
				return "Y";
			case 22:
				return "U";
			case 23:
				return "I";
			case 24:
				return "O";
			case 25:
				return "P";
			case 26:
				return "LEFTBRACE";
			case 27:
				return "RIGHTBRACE";
			case 28:
				return "ENTER";
			case 29:
				return "LEFTCTRL";
			case 30:
				return "A";
			case 31:
				return "S";
			case 32:
				return "D";
			case 33:
				return "F";
			case 34:
				return "G";
			case 35:
				return "H";
			case 36:
				return "J";
			case 37:
				return "K";
			case 38:
				return "L";
			case 39:
				return "SEMICOLON";
			case 40:
				return "APOSTROPHE";
			case 41:
				return "GRAVE";
			case 42:
				return "LEFTSHIFT";
			case 43:
				return "BACKSLASH";
			case 44:
				return "Z";
			case 45:
				return "X";
			case 46:
				return "C";
			case 47:
				return "V";
			case 48:
				return "B";
			case 49:
				return "N";
			case 50:
				return "M";
			case 51:
				return "COMMA";
			case 52:
				return "DOT";
			case 53:
				return "SLASH";
			case 54:
				return "RIGHTSHIFT";
			case 55:
				return "KPASTERISK";
			case 56:
				return "LEFTALT";
			case 57:
				return "SPACE";
			case 58:
				return "CAPSLOCK";
			case 59:
				return "F1";
			case 60:
				return "F2";
			case 61:
				return "F3";
			case 62:
				return "F4";
			case 63:
				return "F5";
			case 64:
				return "F6";
			case 65:
				return "F7";
			case 66:
				return "F8";
			case 67:
				return "F9";
			case 68:
				return "F10";
			case 69:
				return "NUMLOCK";
			case 70:
				return "SCROLLLOCK";
			case 71:
				return "KP7";
			case 72:
				return "KP8";
			case 73:
				return "KP9";
			case 74:
				return "KPMINUS";
			case 75:
				return "KP4";
			case 76:
				return "KP5";
			case 77:
				return "KP6";
			case 78:
				return "KPPLUS";
			case 79:
				return "KP1";
			case 80:
				return "KP2";
			case 81:
				return "KP3";
			case 82:
				return "KP0";
			case 83:
				return "KPDOT";
			case 85:
				return "ZENKAKUHANKAKU";
			case 86:
				return "102ND";
			case 87:
				return "F11";
			case 88:
				return "F12";
			case 89:
				return "RO";
			case 90:
				return "KATAKANA";
			case 91:
				return "HIRAGANA";
			case 92:
				return "HENKAN";
			case 93:
				return "KATAKANAHIRAGANA";
			case 94:
				return "MUHENKAN";
			case 95:
				return "KPJPCOMMA";
			case 96:
				return "KPENTER";
			case 97:
				return "RIGHTCTRL";
			case 98:
				return "KPSLASH";
			case 99:
				return "SYSRQ";
			case 100:
				return "RIGHTALT";
			case 101:
				return "LINEFEED";
			case 102:
				return "HOME";
			case 103:
				return "UP";
			case 104:
				return "PAGEUP";
			case 105:
				return "LEFT";
			case 106:
				return "RIGHT";
			case 107:
				return "END";
			case 108:
				return "DOWN";
			case 109:
				return "PAGEDOWN";
			case 110:
				return "INSERT";
			case 111:
				return "DELETE";
			case 112:
				return "MACRO";
			case 113:
				return "MUTE";
			case 114:
				return "VOLUMEDOWN";
			case 115:
				return "VOLUMEUP";
			case 116:
				return "POWER";
			case 117:
				return "KPEQUAL";
			case 118:
				return "KPPLUSMINUS";
			case 119:
				return "PAUSE";
			case 121:
				return "KPCOMMA";
			case 122:
				return "HANGEUL";
			case 123:
				return "HANJA";
			case 124:
				return "YEN";
			case 125:
				return "LEFTMETA";
			case 126:
				return "RIGHTMETA";
			case 127:
				return "COMPOSE";
			case 128:
				return "STOP";
			case 129:
				return "AGAIN";
			case 130:
				return "PROPS";
			case 131:
				return "UNDO";
			case 132:
				return "FRONT";
			case 133:
				return "COPY";
			case 134:
				return "OPEN";
			case 135:
				return "PASTE";
			case 136:
				return "FIND";
			case 137:
				return "CUT";
			case 138:
				return "HELP";
			case 139:
				return "MENU";
			case 140:
				return "CALC";
			case 141:
				return "SETUP";
			case 142:
				return "SLEEP";
			case 143:
				return "WAKEUP";
			case 144:
				return "FILE";
			case 145:
				return "SENDFILE";
			case 146:
				return "DELETEFILE";
			case 147:
				return "XFER";
			case 148:
				return "PROG1";
			case 149:
				return "PROG2";
			case 150:
				return "WWW";
			case 151:
				return "MSDOS";
			case 152:
				return "COFFEE";
			case 153:
				return "DIRECTION";
			case 154:
				return "CYCLEWINDOWS";
			case 155:
				return "MAIL";
			case 156:
				return "BOOKMARKS";
			case 157:
				return "COMPUTER";
			case 158:
				return "BACK";
			case 159:
				return "FORWARD";
			case 160:
				return "CLOSECD";
			case 161:
				return "EJECTCD";
			case 162:
				return "EJECTCLOSECD";
			case 163:
				return "NEXTSONG";
			case 164:
				return "PLAYPAUSE";
			case 165:
				return "PREVIOUSSONG";
			case 166:
				return "STOPCD";
			case 167:
				return "RECORD";
			case 168:
				return "REWIND";
			case 169:
				return "PHONE";
			case 170:
				return "ISO";
			case 171:
				return "CONFIG";
			case 172:
				return "HOMEPAGE";
			case 173:
				return "REFRESH";
			case 174:
				return "EXIT";
			case 175:
				return "MOVE";
			case 176:
				return "EDIT";
			case 177:
				return "SCROLLUP";
			case 178:
				return "SCROLLDOWN";
			case 179:
				return "KPLEFTPAREN";
			case 180:
				return "KPRIGHTPAREN";
			case 181:
				return "NEW";
			case 182:
				return "REDO";
			case 183:
				return "F13";
			case 184:
				return "F14";
			case 185:
				return "F15";
			case 186:
				return "F16";
			case 187:
				return "F17";
			case 188:
				return "F18";
			case 189:
				return "F19";
			case 190:
				return "F20";
			case 191:
				return "F21";
			case 192:
				return "F22";
			case 193:
				return "F23";
			case 194:
				return "F24";
			case 200:
				return "PLAYCD";
			case 201:
				return "PAUSECD";
			case 202:
				return "PROG3";
			case 203:
				return "PROG4";
			case 205:
				return "SUSPEND";
			case 206:
				return "CLOSE";
			case 207:
				return "PLAY";
			case 208:
				return "FASTFORWARD";
			case 209:
				return "BASSBOOST";
			case 210:
				return "PRINT";
			case 211:
				return "HP";
			case 212:
				return "CAMERA";
			case 213:
				return "SOUND";
			case 214:
				return "QUESTION";
			case 215:
				return "EMAIL";
			case 216:
				return "CHAT";
			case 217:
				return "SEARCH";
			case 218:
				return "CONNECT";
			case 219:
				return "FINANCE";
			case 220:
				return "SPORT";
			case 221:
				return "SHOP";
			case 222:
				return "ALTERASE";
			case 223:
				return "CANCEL";
			case 224:
				return "BRIGHTNESSDOWN";
			case 225:
				return "BRIGHTNESSUP";
			case 226:
				return "MEDIA";
			case 227:
				return "SWITCHVIDEOMODE";
			case 228:
				return "KBDILLUMTOGGLE";
			case 229:
				return "KBDILLUMDOWN";
			case 230:
				return "KBDILLUMUP";
			case 231:
				return "SEND";
			case 232:
				return "REPLY";
			case 233:
				return "FORWARDMAIL";
			case 234:
				return "SAVE";
			case 235:
				return "DOCUMENTS";
			case 236:
				return "BATTERY";
			case 237:
				return "BLUETOOTH";
			case 238:
				return "WLAN";
			case 239:
				return "UWB";
			case 240:
				return "UNKNOWN";
			case 241:
				return "VIDEO_NEXT";
			case 242:
				return "VIDEO_PREV";
			case 243:
				return "BRIGHTNESS_CYCLE";
			case 244:
				return "BRIGHTNESS_ZERO";
			case 245:
				return "DISPLAY_OFF";
			case 246:
				return "WIMAX";
			case 352:
				return "OK";
			case 353:
				return "SELECT";
			case 354:
				return "GOTO";
			case 355:
				return "CLEAR";
			case 356:
				return "POWER2";
			case 357:
				return "OPTION";
			case 358:
				return "INFO";
			case 359:
				return "TIME";
			case 360:
				return "VENDOR";
			case 361:
				return "ARCHIVE";
			case 362:
				return "PROGRAM";
			case 363:
				return "CHANNEL";
			case 364:
				return "FAVORITES";
			case 365:
				return "EPG";
			case 366:
				return "PVR";
			case 367:
				return "MHP";
			case 368:
				return "LANGUAGE";
			case 369:
				return "TITLE";
			case 370:
				return "SUBTITLE";
			case 371:
				return "ANGLE";
			case 372:
				return "ZOOM";
			case 373:
				return "MODE";
			case 374:
				return "KEYBOARD";
			case 375:
				return "SCREEN";
			case 376:
				return "PC";
			case 377:
				return "TV";
			case 378:
				return "TV2";
			case 379:
				return "VCR";
			case 380:
				return "VCR2";
			case 381:
				return "SAT";
			case 382:
				return "SAT2";
			case 383:
				return "CD";
			case 384:
				return "TAPE";
			case 385:
				return "RADIO";
			case 386:
				return "TUNER";
			case 387:
				return "PLAYER";
			case 388:
				return "TEXT";
			case 389:
				return "DVD";
			case 390:
				return "AUX";
			case 391:
				return "MP3";
			case 392:
				return "AUDIO";
			case 393:
				return "VIDEO";
			case 394:
				return "DIRECTORY";
			case 395:
				return "LIST";
			case 396:
				return "MEMO";
			case 397:
				return "CALENDAR";
			case 398:
				return "RED";
			case 399:
				return "GREEN";
			case 400:
				return "YELLOW";
			case 401:
				return "BLUE";
			case 402:
				return "CHANNELUP";
			case 403:
				return "CHANNELDOWN";
			case 404:
				return "FIRST";
			case 405:
				return "LAST";
			case 406:
				return "AB";
			case 407:
				return "NEXT";
			case 408:
				return "RESTART";
			case 409:
				return "SLOW";
			case 410:
				return "SHUFFLE";
			case 411:
				return "BREAK";
			case 412:
				return "PREVIOUS";
			case 413:
				return "DIGITS";
			case 414:
				return "TEEN";
			case 415:
				return "TWEN";
			case 416:
				return "VIDEOPHONE";
			case 417:
				return "GAMES";
			case 418:
				return "ZOOMIN";
			case 419:
				return "ZOOMOUT";
			case 420:
				return "ZOOMRESET";
			case 421:
				return "WORDPROCESSOR";
			case 422:
				return "EDITOR";
			case 423:
				return "SPREADSHEET";
			case 424:
				return "GRAPHICSEDITOR";
			case 425:
				return "PRESENTATION";
			case 426:
				return "DATABASE";
			case 427:
				return "NEWS";
			case 428:
				return "VOICEMAIL";
			case 429:
				return "ADDRESSBOOK";
			case 430:
				return "MESSENGER";
			case 431:
				return "DISPLAYTOGGLE";
			case 432:
				return "SPELLCHECK";
			case 433:
				return "LOGOFF";
			case 434:
				return "DOLLAR";
			case 435:
				return "EURO";
			case 436:
				return "FRAMEBACK";
			case 437:
				return "FRAMEFORWARD";
			case 438:
				return "CONTEXT_MENU";
			case 439:
				return "MEDIA_REPEAT";
			case 448:
				return "DEL_EOL";
			case 449:
				return "DEL_EOS";
			case 450:
				return "INS_LINE";
			case 451:
				return "DEL_LINE";
			case 464:
				return "FN";
			case 465:
				return "FN_ESC";
			case 466:
				return "FN_F1";
			case 467:
				return "FN_F2";
			case 468:
				return "FN_F3";
			case 469:
				return "FN_F4";
			case 470:
				return "FN_F5";
			case 471:
				return "FN_F6";
			case 472:
				return "FN_F7";
			case 473:
				return "FN_F8";
			case 474:
				return "FN_F9";
			case 475:
				return "FN_F10";
			case 476:
				return "FN_F11";
			case 477:
				return "FN_F12";
			case 478:
				return "FN_1";
			case 479:
				return "FN_2";
			case 480:
				return "FN_D";
			case 481:
				return "FN_E";
			case 482:
				return "FN_F";
			case 483:
				return "FN_S";
			case 484:
				return "FN_B";
			case 497:
				return "BRL_DOT1";
			case 498:
				return "BRL_DOT2";
			case 499:
				return "BRL_DOT3";
			case 500:
				return "BRL_DOT4";
			case 501:
				return "BRL_DOT5";
			case 502:
				return "BRL_DOT6";
			case 503:
				return "BRL_DOT7";
			case 504:
				return "BRL_DOT8";
			case 505:
				return "BRL_DOT9";
			case 506:
				return "BRL_DOT10";
			case 511:
				return "MAX";
			default:
				if (! ret_val) ret_val = malloc (12);
				sprintf(ret_val, "KEY_%d", ev.code);
				return ret_val;

		}

	}

	return NULL;

}

